apache/metron
# METRON-1367 Stellar should have some instrumentation of function calls

This PR introduces the ability through a new `magic` command for a stellar shell user to view the timing information for the last statement the shell executed.  The information provided is the total execution time, and the execution time per stellar function, and is tagged when creating the timings.
These tags can be passed to ~~%timing~~ %time to filter the output.  In other words if you call

~~`%timing FUNCTION`~~ %time FUNCTION  you will only the total time, and any timings that have been tagged as FUNCTION

At this point, only the stellar function and lambda executions have been timed, but it is possible in the future to add timing to other parts of stellar.

For example:

```bash
Stellar, Go!
Please note that functions are loading lazily in the background and will be unavailable until loaded fully.
[Stellar]>>> Functions loaded, you may refer to functions now...

[Stellar]>>> hex:="91IMOR3F41BMUSJCCG======"
[Stellar]>>> if( STARTS_WITH(TO_UPPER(DECODE(hex,'BASE32HEX')), "HELLO")) THEN TO_LOWER("OK") ELSE TO_LOWER("FFFF")
ok
[Stellar]>>> %time
->execute : 113ms : 113017132
-->if( STARTS_WITH(TO_UPPER(DECODE(hex,'BASE32HEX')), "HELLO")) THEN TO_LOWER("OK") ELSE TO_LOWER("FFFF") : 112ms : 112998881ns
--->DECODE : 8ms : 8572968ns
--->TO_UPPER : 0ms : 20018ns
--->STARTS_WITH : 0ms : 30586ns
--->TO_LOWER : 0ms : 19085ns
```
or more complex, showing tags and filtering

```bash
[Stellar]>>> foo:=1
[Stellar]>>> match { foo == 0 => ()-> false, foo == 1 => ()-> TO_UPPER("true"), default => ()-> false }
TRUE
[Stellar]>>> %time
->execute[] : 4242ms : 4242237640ns
-->match { foo == 0 => ()-> false, foo == 1 => ()-> TO_UPPER("true"), default => ()-> false }[] : 84ms : 84362570ns
--->lambda[LAMBDA] : 1ms : 1265895ns
---->TO_UPPER[FUNCTION] : 0ms : 570672ns

[Stellar]>>> %time FUNCTION
->execute[] : 7594ms : 7594583655ns
-->match { foo == 0 => ()-> false, foo == 1 => ()-> TO_UPPER("true"), default => ()-> false }[] : 84ms : 84362570ns
---->TO_UPPER[FUNCTION] : 0ms : 570672ns

[Stellar]>>> %time LAMBDA
->execute[] : 14374ms : 14374556116ns
-->match { foo == 0 => ()-> false, foo == 1 => ()-> TO_UPPER("true"), default => ()-> false }[] : 84ms : 84362570ns
--->lambda[LAMBDA] : 1ms : 1265895ns
```

## StackWatch
A set of utility classes under the new package stellar.common.timing have been added.  These provide the `StackWatch` functionality.

StackWatch provides an abstraction over the Apache Commons StopWatch class that allows callers to 
create multiple named and possibly nested timing operations.

Since Stellar operations are executed in a queue and in isolation ( even a Stellar statement that is nested  such as  TO_UPPER(TO_STRING()) is executed as  TO_STRING(), TO_UPPER, the benefit of the support for nested timing operations is not utilized.  If the function is part of a LambdaExpression, you will however see the nesting, as shown in the example above.

This class may be more generally useful to this and other projects, but I am not sure where it would live since we wouldn't want it in common.

StackWatch uses a combination of `Deque` and a custom Tree implementation to create, start and end timing operations.

A `Visitor` pattern is also implemented to allow for retrieving the results after the completion of the operation.

See the StackWatch tests for examples of nested function timings.

The StackWatch is optionally available from the Stellar `Context` object.

The shell stores the watch from the previous run, while clearing it from the context.  This allows for multiple %timing calls to be executed ( multiple visits ), each possibly with different parameters.

> NOTE: I ~~am going to see about submitting~~ have submitted  https://github.com/apache/commons-lang/pull/311 to Apache Commons Lang as well, if that happens, we would switch over to the official implementation.

## Testing

To test:

- Run the Stellar Shell, either from full_dev or using mvn exec ( here run from metron-stellar/stellar-common after a mvn package )
```
> % mvn exec:java  -Dexec.mainClass="org.apache.metron.stellar.common.shell.StellarShell"
```

- run some stellar statements
- run the %time 
- run the %time FUNCTION 
- run the %time LAMBDA
- in full dev verify that data is processed normally through storm, and that it shows in kibana





### For code changes:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```
